### PR TITLE
Document the design principles behind AppAuth

### PR DESCRIPTION
**Pull request by WilliamDenniss**
Originally drafted June 1, 2018.

### CODE REVIEW

File: DESIGN.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+requests are thoroughly reviewed for style, the quality of the API, and future maintainability.
+
+### Review Time
+
+Expect review periods of 6 months to a year for integrating major new in-scope features. Our
+priority as previously stated is stability and quality, not adding features as quickly as possible.
+While you work with us to integrate your feature, we highly encourage you to maintain your own fork
+so you and others can use the feature immediately – and gain valuable implementation experience.
+
+
```
Comment (petea) [inline DESIGN.md:122]: Drop extra empty lines.

File: DESIGN.md [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+these extension points. For example, simple additional parameters can be passed in the
+“additionalParameters” directory, and the TokenRequest can support other grant types. 
+
+By way of example, the tvOS device flow support (which may actually be considered in-scope one day)
+was initially implemented on top of these extension points without needing to change AppAuth.
+
+### Provider-specific workarounds, hacks or features
+
+AppAuth implements the pure authentication and authorization standards. Where these standards are
+clear in their meaning, errors in provider implementations are supported or worked around by
```
[the remark this answers is not in the document]
Comment (WilliamDenniss) [reply DESIGN.md]: Yes, fixed.

File: DESIGN.md [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+Likewise, some specifications document several different sub-protocols. We don’t aim to implement
+every single branch of the spec.
+
+### Non-standard protocols and parameters
+
+Non-standard protocols and parameters are not supported by AppAuth.  However, AppAuth, like the
+specs it implements, is largely extensible and you may be able to achieve what you need through
+these extension points. For example, simple additional parameters can be passed in the
+“additionalParameters” directory, and the TokenRequest can support other grant types.
```
[the remark this answers is not in the document]
Comment (WilliamDenniss) [reply DESIGN.md]: Thanks, fixed!

File: DESIGN.md (hunk 1 above)
Comment (WilliamDenniss) [reply DESIGN.md]: Done

**Pull request merged**